### PR TITLE
add vector-to-array clang_delta transform

### PR DESCRIPTION
**Pull request by oriceemple**
A new clang_delta transform to replace vectors with arrays. We noticed that doing this manually helped us reduce in many cases.

### CODE REVIEW

**Comment by regehr** [comment]
This looks great but Yang owns clang_delta and I'd like him to give this a once-over before we merge.

**Comment by chenyang78** [comment]
@oriceemple Thanks for the patch! I will take a look at it soon.

File: clang_delta/VectorToArray.cpp [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+#define ARRAY_SUFFIX "[10]"
+
+using namespace clang;
+
+static const char *DescriptionMsg =
+    "Try replacing vectors with arrays. It transforms the following code: \n\
+  vector<int> v' \n\
+to \n\
+  int v[8]\n";
```
Comment (chenyang78) [inline clang_delta/VectorToArray.cpp]: To match ```ARRAY_SUFFIX```,  change it to```int v[10]```?

File: clang_delta/VectorToArray.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+static RegisterTransformation<VectorToArray> Trans("vector-to-array",
+                                                   DescriptionMsg);
+
+class VectorToArrayCollectionVisitor
+    : public RecursiveASTVisitor<VectorToArrayCollectionVisitor> {
+public:
+  explicit VectorToArrayCollectionVisitor(VectorToArray *instance)
```
Comment (chenyang78) [inline clang_delta/VectorToArray.cpp]: ```s/instance/Instance/``` ?

File: clang_delta/VectorToArray.cpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+      : Consumer(instance) {}
+
+  bool VisitNamedDecl(NamedDecl *D) { return true; }
+
+  bool VisitClassTemplateDecl(ClassTemplateDecl *D) {
+    if (Consumer->isInIncludedFile(D))
+      return true;
+    auto *ns = dyn_cast<NamespaceDecl>(D->getDeclContext());
```
Comment (chenyang78) [inline clang_delta/VectorToArray.cpp]: ```ns``` => ```NS```?

File: clang_delta/VectorToArray.cpp [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+  //     return true;
+
+  //   QualType QT = VD->getType();
+  //   const Type *T = QT->getTypePtr();
+  //   const CXXRecordDecl *RD = T->getAsCXXRecordDecl();
+  //   if (RD == nullptr)
+  //     return true;
+  //   continue from here
+
+  // }
```
Comment (chenyang78) [inline clang_delta/VectorToArray.cpp]: Maybe delete the entire commented code?

File: clang_delta/VectorToArray.cpp [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+private:
+  VectorToArray *Consumer;
+};
+
+class VectorToArrayRewriteVisitor
+    : public RecursiveASTVisitor<VectorToArrayRewriteVisitor> {
+private:
+  const std::string getVectorElemTypeName(VarDecl *VD) {
```
Comment (chenyang78) [inline clang_delta/VectorToArray.cpp]: Not sure if we need to return ```const std::string```.

File: clang_delta/VectorToArray.cpp [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+private:
+  VectorToArray *Consumer;
+};
+
+void VectorToArray::Initialize(ASTContext &ctx) {
+  Transformation::Initialize(ctx);
+  CollectionVisitor = new VectorToArrayCollectionVisitor(this);
+  RewriteVisitor = new VectorToArrayRewriteVisitor(this);
+}
```
Comment (chenyang78) [inline clang_delta/VectorToArray.cpp]: ```delete CollectionVisitor``` and ```RewriteVisitor``` in ```~VectorArray```?

File: clang_delta/VectorToArray.cpp [hunk 7]
```diff
@@ -0,0 +1,188 @@
+
```
Comment (chenyang78) [inline clang_delta/VectorToArray.cpp]: Need copyright here.

File: clang_delta/VectorToArray.h [hunk 8]
```diff
@@ -0,0 +1,35 @@
+#ifndef VECTOR_TO_ARRAY_H
```
Comment (chenyang78) [inline clang_delta/VectorToArray.h]: Copyright info.

**Comment by chenyang78** [comment]
@oriceemple Can you also update ```Makefile.am``` and ```Makefile.in```? Again, thanks for the patch!

**Comment by oriceemple** [comment]
I have created a fixed version, check out https://github.com/csmith-project/creduce/pull/108

**Pull request closed without merging**